kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java

```diff
@@ -1068,61 +1084,51 @@ public Boolean isReady() {
     return i instanceof HasMetadata && Readiness.isReady((HasMetadata)i);
   }
 
-  protected T waitUntilExists(long amount, TimeUnit timeUnit) throws InterruptedException {
-    return waitUntilCondition(Objects::nonNull, amount, timeUnit);
-  }
-
   @Override
   public T waitUntilReady(long amount, TimeUnit timeUnit) throws InterruptedException {
-
-    long timeoutInNanos = timeUnit.toNanos(amount);
-    long end = System.nanoTime() + timeoutInNanos;
-
-    while (System.nanoTime() < end) {
-      T item = fromServer().get();
-      try {
-        if (Readiness.isReady((HasMetadata) item)) {
-          return item;
-        }
-
-        Thread.sleep(500);
-      } catch (IllegalArgumentException illegalArgumentException) {
-        // This might be thrown if Resource passed doesn't comply with concept of "readiness"
-        throw illegalArgumentException;
-      }
-    }
-
-    T item = fromServer().get();
-    if (Readiness.isReady((HasMetadata) item)) {
-      return item;
-    }
-
-    throw new IllegalStateException(type.getSimpleName() + " with name:[" + name + "] in namespace:[" + namespace + "] not ready!");
+    return waitUntilCondition(resource -> Objects.nonNull(resource) && Readiness.isReady(resource), amount, timeUnit);
   }
 
   @Override
   public T waitUntilCondition(Predicate<T> condition, long amount, TimeUnit timeUnit)
     throws InterruptedException {
+    return waitUntilConditionWithRetries(condition, timeUnit.toMillis(amount), watchRetryInitialBackoffMillis);
+  }
 
-    long timeoutInMillis = timeUnit.toNanos(amount);
-
-    long end = System.nanoTime() + timeoutInMillis;
-    while (System.nanoTime() < end) {
-      T item = get();
-      if (condition.test(item)) {
-        return item;
-      }
-
-      // in the future, this should probably be more intelligent
-      Thread.sleep(500);
-    }
+  private T waitUntilConditionWithRetries(Predicate<T> condition, long timeoutMillis, long backoffMillis)
+    throws InterruptedException {
 
-    T item = get();
+    T item = fromServer().get();
     if (condition.test(item)) {
       return item;
     }
 
-    throw new IllegalArgumentException(type.getSimpleName() + " with name:[" + name + "] in namespace:[" + namespace + "] not found!");
+    long end = System.currentTimeMillis() + timeoutMillis;
```

Comment:
technically we should use `nanoTime` for this, `currenTimeMillis` is not monotonic and so it doesn't really work for timing operations like this.